step/step.go

```diff
@@ -0,0 +1,566 @@
+			s.logger.Warnf("xcode build log:\n%s", buildLog)
+			s.logger.Errorf("xcode build failed with error: %s", err)
+			return result, -1, err
+		}
+	}
+
+	// Run test
+	tempDir, err := ioutil.TempDir("", "XCUITestOutput")
+	if err != nil {
+		return result, -1, fmt.Errorf("could not create test output temporary directory: %s", err)
+	}
+	xcresultPath := path.Join(tempDir, "Test.xcresult")
+
+	testParams := xcodebuild.TestParams{
+		BuildParams:                    buildParams,
+		TestPlan:                       cfg.TestPlan,
+		TestOutputDir:                  xcresultPath,
+		TestRepetitionMode:             cfg.TestRepetitionMode,
+		MaximumTestRepetitions:         cfg.MaximumTestRepetitions,
+		RelaunchTestsForEachRepetition: cfg.RelaunchTestForEachRepetition,
+		BuildBeforeTest:                cfg.BuildBeforeTesting,
+		GenerateCodeCoverage:           cfg.GenerateCodeCoverageFiles,
+		RetryTestsOnFailure:            cfg.RetryTestsOnFailure,
+		AdditionalOptions:              cfg.XcodebuildTestOptions,
+	}
+
+	if cfg.IsSingleBuild {
+		testParams.CleanBuild = cfg.IsCleanBuild
+	}
+
+	var swiftPackagesPath string
+	if cfg.XcodeMajorVersion >= 11 {
+		var err error
+		swiftPackagesPath, err = s.cache.SwiftPackagesPath(cfg.ProjectPath)
+		if err != nil {
+			return result, -1, fmt.Errorf("failed to get Swift Packages path, error: %s", err)
+		}
+	}
+
+	params := xcodebuild.TestRunParams{
+		BuildTestParams:                    testParams,
+		OutputTool:                         cfg.OutputTool,
+		XcprettyOptions:                    cfg.XcprettyOptions,
+		RetryOnTestRunnerError:             true,
+		RetryOnSwiftPackageResolutionError: true,
+		SwiftPackagesPath:                  swiftPackagesPath,
+		XcodeMajorVersion:                  cfg.XcodeMajorVersion,
+	}
+	testLog, exitCode, testErr := s.xcodebuild.RunTest(params)
+	result.XcresultPath = xcresultPath
+	result.XcodebuildTestLog = testLog
+
+	if testErr != nil || cfg.OutputTool == xcodebuild.XcodebuildTool {
+		printLastLinesOfXcodebuildTestLog(testLog, testErr == nil)
+	}
+
+	return result, exitCode, testErr
+}
+
+func (s XcodeTestRunner) teardownSimulator(simulatorID string, simulatorDebug exportCondition, isSimulatorBooted bool, testErr error) string {
+	var simulatorDiagnosticsPath string
+
+	if simulatorDebug == always || (simulatorDebug == onFailure && testErr != nil) {
+		s.logger.Println()
+		s.logger.Infof("Collecting Simulator diagnostics")
+
+		diagnosticsPath, err := s.simulatorManager.SimulatorCollectDiagnostics()
+		if err != nil {
+			s.logger.Warnf("%v", err)
+		} else {
+			s.logger.Donef("Simulator diagnostics are available as an artifact (%s)", diagnosticsPath)
+			simulatorDiagnosticsPath = diagnosticsPath
+		}
+	}
+
+	// Shut down the simulator if it was started by the step for diagnostic logs.
+	if !isSimulatorBooted && simulatorDebug != never {
+		if err := s.simulatorManager.SimulatorShutdown(simulatorID); err != nil {
```

Comment:
Shouldn't we retry cleaning up when we fail? When we'll have proper analytics it would be nice to see the frequency of "edge" cases like this.

Reply:
It would be great, I tried to avoid functional changes in this PR, so it stayed as it was: https://github.com/bitrise-steplib/steps-xcode-test/blob/master/main.go#L479